pcap/pcapio/ngread.go

```diff
@@ -0,0 +1,310 @@
+// packet types aren't supported yet (we presume thiis type of trace is rare and
+// does not fit our use case here as these traces do not include capture timestamps
+// annd the point here is to pull our ranges of packets from a large pcap based
+// on timestamp).  We also do not support the original deprecated PCAP-NG packet
+// format but could add support if users request this (it would only be because
+// old pcaps with this deprecated format are sitting around).
+func (r *NgReader) Packet(block []byte) ([]byte, nano.Ts, layers.LinkType) {
+	packet, ifno := r.parsePacket(block)
+	if packet == nil {
+		return nil, 0, 0
+	}
+	ts := uint64(r.getUint32(block[12:16]))<<32 | uint64(r.getUint32(block[16:20]))
+	t := time.Unix(r.convertTime(ifno, ts)).UTC()
+	return packet, nano.TimeToTs(t), r.ifaces[ifno].LinkType
+}
+
+func (r *NgReader) Offset() uint64 {
+	return r.offset
+}
+
+// The following functions make the binary.* functions inlineable (except for getUint64, which is too big, but not in any hot path anyway)
+// Compared to storing binary.*Endian in a binary.ByteOrder this shaves off about 20% for (ZeroCopy)ReadPacketData, which is caused by the needed itab lookup + indirect go call
+func (r *NgReader) getUint16(buffer []byte) uint16 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint16(buffer)
+	}
+	return binary.LittleEndian.Uint16(buffer)
+}
+
+func (r *NgReader) getUint32(buffer []byte) uint32 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint32(buffer)
+	}
+	return binary.LittleEndian.Uint32(buffer)
+}
+
+func (r *NgReader) getUint64(buffer []byte) uint64 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint64(buffer)
+	}
+	return binary.LittleEndian.Uint64(buffer)
+}
+
+// Now the pcapng implementation
+
+// readBlock reads a the blocktype and length from the file.
+// If the type is a section header, endianess is also read.
+func (r *NgReader) readBlock() (ngBlockType, []byte, error) {
+	hdr, err := r.Peek(8)
+	if err != nil {
+		return 0, nil, err
+	}
+	typ := ngBlockType(r.getUint32(hdr[:4]))
+	length := r.getUint32(hdr[4:8])
+	if length == 0 {
+		// avoid infinite loop for bad input
+		return 0, nil, errors.New("zero-length pcap-ng block")
+	}
+	// The next part is a bit messed up since a section header could change the endianess...
+	// So first read then length just into a buffer, check if its a section header and then do the endianess part...
+	//XXX move this above
+	b, err := r.Reader.Read(int(length))
+	if uint32(len(b)) < length {
+		return 0, nil, errors.New("truncated pcap-ng block")
+	}
+	if r.getUint32(b[length-4:length]) != uint32(length) {
+		return 0, nil, errors.New("pcap-ng trailer length mismatch")
+	}
+	return typ, b, err
+}
+
+func (r *NgReader) parseSectionHeader(b []byte) error {
+	if len(b) < 12 {
+		return ErrCorruptPcap
+	}
+	if binary.BigEndian.Uint32(b[8:12]) == ngByteOrderMagic {
+		r.bigEndian = true
+	} else if binary.LittleEndian.Uint32(b[8:12]) == ngByteOrderMagic {
+		r.bigEndian = false
```

Comment:
nit: Instead of this why not replace `bigEndian bool` with `byteOrder binary.ByteOrder` and do:
```golang
if binary.BigEndian.Uint32(b[8:12]) == ngByteOrderMagic {
	r.byteOrder = binary.BigEndian
} else if binary.LittleEndian.Uint32(b[8:12]) == ngByteOrderMagic {
	r.byteOrder = binary.LittleEndian
```
This way you can get rid of your `getUint64/32/16` helpers.

Reply:
How about later?  This is all gopacket logic that I was just hoping would work for us.

Reply:
ok